### ISSUE TITLE
Allow embedded neural network evaluation to have side effects

### ISSUE THREAD

**Issue by jrmaddison** [jrmaddison/bt_ocean#37]
Required e.g. for batch normalization

**Comment by jrmaddison** [comment]
The simplest way to do this might be to register Keras `Layer`s as pytree nodes, repurposing Keras serialization for the flatten and unflatten, and then swap the `vmap` and `fori_loop` used to evaluate the `Dynamics` layer to that the embedded network operates on the whole batch. It might alternatively be possible to let `Dynamics` be a `keras.layers.RNN`.

**Comment by jrmaddison** [comment]
The correct way to do this is via `stateless_call`.

**Comment by jrmaddison** [comment]
There are a few ways stateful updates might be introduced by adding carry arguments. This will need some thought when there is a case which uses it, but for now #54 detects the issue.
